arch/arm/core/cortex_m/arm-m-switch.c

```diff
@@ -0,0 +1,437 @@
+ * size of this struct, it just wants to be have compiler-visible
+ * offsets for in-place copies.
+ */
+/* clang-format off */
+#define FRAMESZ (4 + MAX(sizeof(struct z_frame_fpu), sizeof(struct hw_frame_align_fpu)))
+#define PAD(T)  char pad_##T[FRAMESZ - sizeof(struct T)]
+union frame {
+	struct { PAD(hw_frame_base);      struct hw_frame_base hw;          };
+	struct { PAD(hw_frame_fpu);       struct hw_frame_fpu hwfp;         };
+	struct { PAD(hw_frame_align);     struct hw_frame_align hw_a;       };
+	struct { PAD(hw_frame_align_fpu); struct hw_frame_align_fpu hwfp_a; };
+	struct { PAD(z_frame);            struct z_frame z;                 };
+	struct { PAD(z_frame_fpu);        struct z_frame_fpu zfp;           };
+};
+/* clang-format on */
+
+/* Validate the structs are correctly top-aligned */
+#define FRAME_FIELD_END(F) ((void *)&(&(((union frame *)0)->F))[1])
+BUILD_ASSERT(FRAME_FIELD_END(hw) == FRAME_FIELD_END(hwfp));
+BUILD_ASSERT(FRAME_FIELD_END(hw) == FRAME_FIELD_END(hw_a));
+BUILD_ASSERT(FRAME_FIELD_END(hw) == FRAME_FIELD_END(hwfp_a));
+BUILD_ASSERT(FRAME_FIELD_END(hw) == FRAME_FIELD_END(z));
+BUILD_ASSERT(FRAME_FIELD_END(hw) == FRAME_FIELD_END(zfp));
+
+#ifdef CONFIG_FPU_SHARING
+uint32_t arm_m_switch_stack_buffer = sizeof(struct z_frame_fpu) - sizeof(struct hw_frame_base);
+#else
+uint32_t arm_m_switch_stack_buffer = sizeof(struct z_frame) - sizeof(struct hw_frame_base);
+#endif
+
+/* Global pointers to the frame locations for the callee-saved
+ * registers.  Set in arm_m_must_switch(), and used by the fixup
+ * assembly in arm_m_exc_exit.
+ */
+struct {
+	void *out, *in;
+} arm_m_cs_ptrs;
+
+#ifdef CONFIG_LTO
+/* Toolchain workaround: when building with LTO, gcc seems unable to
+ * notice the reference to arm_m_cs_ptrs in the assembly for
+ * arm_m_exc_exit below, and drops the symbol before the final link.
+ * Use this global to store a pointer in arm_m_new_stack(), wasting a
+ * few bytes of code & data.
+ */
+void *arm_m_lto_dummy;
+#endif
+
+/* Unit test hook, unused in production */
+void *arm_m_last_switch_handle;
+
+/* Global holder for the location of the saved LR in the entry frame. */
+uint32_t *arm_m_exc_lr_ptr;
+
+/* Dummy used in arch_switch() when USERSPACE=y */
+uint32_t arm_m_switch_control;
+
+/* clang-format off */
+/* Emits an in-place copy from a hw_frame_base to a switch_frame */
+#define HW_TO_SWITCH(hw, sw) do {					\
+	uint32_t r0 = hw.r0, r1 = hw.r1, r2 = hw.r2, r3 = hw.r3;        \
+	uint32_t r12 = hw.r12, lr = hw.lr, pc = hw.pc, apsr = hw.apsr;  \
+	pc |= 1; /* thumb bit! */                                       \
+	sw.r0 = r0; sw.r1 = r1; sw.r2 = r2; sw.r3 = r3;                 \
+	sw.r12 = r12; sw.lr = lr; sw.pc = pc; sw.apsr = apsr;           \
+} while (false)
+
+/* Emits an in-place copy from a switch_frame to a synth_frame */
+#define SWITCH_TO_SYNTH(sw, syn) do {					\
+	struct synth_frame syntmp = {                                   \
```

Comment:
Note to myself here: somewhere I dropped a change.  This is using my original "copy a temp struct" code (which the compiler had trouble with and was emitting memcpy() calls in some situations!), where I later ported it to a bunch of individual variables which it did much better optimizing.  You can see it above in HW_TO_SWITCH().  But somehow I only committed one half!?  Should re-fix, this is probably good for a dozen cycles or so.